src/Bicep.Core/TypeSystem/DeployTimeConstantVisitor.cs

```diff
@@ -0,0 +1,178 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+using Bicep.Core.Diagnostics;
+using Bicep.Core.Semantics;
+using Bicep.Core.Syntax;
+using System;
+
+namespace Bicep.Core.TypeSystem
+{
+    /// <summary>
+    /// Visitor used to collect errors caused by property assignments to run-time values when that is not allowed
+    /// </summary>
+    public sealed class DeployTimeConstantVisitor : SyntaxVisitor
+    {
+
+        private readonly SemanticModel model;
+        private readonly IDiagnosticWriter diagnosticWriter;
+
+        private ObjectType? bodyObj;
+
+        // Since we collect errors top down, we will collect and overwrite this variable and emit it in the end
+        private SyntaxBase? errorSyntax;
+        private string? currentSymbol;
+
+
+        private DeployTimeConstantVisitor(SemanticModel model, IDiagnosticWriter diagnosticWriter)
+        {
+            this.model = model;
+            this.diagnosticWriter = diagnosticWriter;
+        }
+
+        // entry point for this visitor. We only iterate through modules and resources
+        public static void ValidateDeployTimeConstants(SemanticModel model, IDiagnosticWriter diagnosticWriter)
+        {
+            var deploymentTimeConstantVisitor = new DeployTimeConstantVisitor(model, diagnosticWriter);
+            foreach (var declaredSymbol in model.Root.ResourceDeclarations)
+            {
+                deploymentTimeConstantVisitor.Visit(declaredSymbol.DeclaringSyntax);
+            }
+            foreach (var declaredSymbol in model.Root.ModuleDeclarations)
+            {
+                deploymentTimeConstantVisitor.Visit(declaredSymbol.DeclaringSyntax);
+            }
+        }
+
+        public override void VisitResourceDeclarationSyntax(ResourceDeclarationSyntax syntax)
+        {
+            if (model.GetSymbolInfo(syntax) is ResourceSymbol resourceSymbol &&
```

Comment:
```csharp
if (model.GetTypeInfo(syntax) is ResourceType resourceType &&
     resourceType.Body is ObjectType bodyObj)
```

Reply:
![image](https://user-images.githubusercontent.com/3432839/101962526-a3f2dd80-3bc9-11eb-981d-79d985ed34b0.png)
This doesn't seem to have the same effect

Reply:
I suspect the difference is that the top is showing you the 'declared' type (think of this as the type definition fro swagger), and the bottom is showing you the 'assigned' type (think of this as the type definition but filtered based on the actual syntax). Barring a few special cases, I would expect the assigned type to be the type to use for this sort of analysis. Is there a reason you need to use the declared type?

Reply:
We need the declared type to be able to distinguish all `DeployTimeConstant` fields of the referenced object.

Reply:
Will circle back at this in https://github.com/Azure/bicep/issues/1177